ddp-workspace/projects/ddp-sdk/src/lib/components/activityForm/answers/activity-composite-answer/activityCompositeAnswer.component.ts

```diff
@@ -42,36 +78,80 @@ export class ActivityCompositeAnswer implements OnChanges {
-                    });
-
-                    const blankRow: ActivityQuestionBlock<any>[] = this.block.children.map((questionBlock: ActivityQuestionBlock<any>) =>
-                        this.buildBlockForChildQuestion(questionBlock, null, this.block.shown));
-
-                    // If a row in Composite block has 2 and more questions,
-                    // and a user answered only on the first question and reloaded the page
-                    // the backend returns an array of answers only with 1 item,
-                    // so when the component will build rows, we will miss some blocks
-                    // example: https://broadinstitute.atlassian.net/browse/DDP-4536;
-                    // method below looks which blocks were missed and add them
-                    // Important: if a user answered only on the last question, backend returns the correct array of answers
-                    this.childQuestionBlocks = questionsRows.map((currentRow: ActivityQuestionBlock<any>[]) => {
-                        if (currentRow.length !== blankRow.length) {
-                            const missedQuestions = blankRow.slice(currentRow.length);
-                            currentRow.push(...missedQuestions);
-                        }
-                        return currentRow;
-                    });
+                    this.childQuestionBlocks = this.rebuildChildQuestions(newBlock, newAnswers);
+                }
+            }
+        }
+    }
+
+    private updateEquationQuestions(equations: AnswerResponseEquation[] = []): void {
+        // will keep here equations values which were updated
+        const equationsToUpdate: {[stableId: string]: DecimalAnswer[]} = {};
+
+        const childEquations = this.block.children
+            .filter(question => question.questionType === QuestionType.Equation) as  ActivityEquationQuestionBlock[];
+
+        // add equations answers to the parent composite question answer
+        for (const equation of equations) {
+            for (const childEquation of childEquations) {
+                if (childEquation.stableId === equation.stableId) {
+                    equationsToUpdate[childEquation.stableId] = equation.values;
+
+                    const prevAnswer: AnswerContainer[][] = this.block.answer;
+                    const newAnswer: AnswerContainer[][] = prevAnswer.map((answerRow: AnswerContainer[], index) =>
+                        // eslint-disable-next-line arrow-body-style
+                        answerRow.map((answer: AnswerContainer) => {
+                            return (answer.stableId === childEquation.stableId) ? {
+                                ...answer,
+                                value: [equation.values[index]]
+                            } : answer;
+                        })
+                    );
+                    this.block.setAnswer(newAnswer, false);
```

Comment:
Would be desirable that composite questions don't know about equations either.
This probably can be achieved if each equation.component and/or equation.block knows that it is part of a composite answer AND what its index position it is in.
With this information, by subscribing to the Patch response, the equation component can pick out its own answer ("My stableid is X, my index pos is Y, so this must be my Equation value from the response").